File: app/src/main/java/org/rul/nutrimake/dao/ClienteDao.java
package org.rul.nutrimake.dao;

import android.arch.persistence.room.Dao;
import android.arch.persistence.room.Delete;
import android.arch.persistence.room.Insert;
import android.arch.persistence.room.Query;

import org.rul.nutrimake.model.Cliente;

import java.util.List;

import static android.arch.persistence.room.OnConflictStrategy.IGNORE;

/**
 * Created by Rul on 02/11/2017.
 */
@Dao
public interface ClienteDao {
    
    //Nuevo Cliente    
    @Insert(onConflict = REPLACE) //Modificar datos Cliente
    void insert(Cliente cliente);
    
    //Lista Clientes
    @Query("select * from cliente")
    List<Cliente> findAll();
    
    //Recuperar un Cliente por id
    @Query("select * from cliente where id = :id")
    Cliente findById(int id);

    //Recuperar un Cliente por nombre y/o apellidos
    @Query("select * from cliente where nombre = :nombre and apellidos = :apellidos")
    List<Cliente> findByNameAndLastName(String nombre, String apellidos);
    
    //Eliminar un Cliente
    @Delete
    void delete(Cliente cliente);

    ////////////////////////////////
    // Otros metodos para pruebas //
    ////////////////////////////////
    
    @Insert(onConflict = IGNORE)
    void insertOrReplaceClientes(Cliente... clientes);

    @Query("delete from cliente where nombre like :badName OR apellidos like :badName")
    int deleteClientesByNombre(String badName);

    @Delete
    void deleteClientes(Cliente cliente1, Cliente cliente2);

    @Query("SELECT * FROM Cliente WHERE :edad == :edad") // TODO: Fix this!
    List<Cliente> findYoungerThan(int edad);

    @Query("SELECT * FROM Cliente WHERE edad < :edad")
    List<Cliente> findYoungerThanSolution(int edad);

    @Query("DELETE FROM Cliente")
    void deleteAll();
    
}

File: app/src/main/java/org/rul/nutrimake/dao/ClienteDao.kt
package org.rul.nutrimake.dao

import android.arch.lifecycle.LiveData
import android.arch.persistence.room.Dao
import android.arch.persistence.room.Delete
import android.arch.persistence.room.Insert
import android.arch.persistence.room.Query
import android.database.Cursor
import io.reactivex.Flowable
import org.rul.nutrimake.model.Cliente
import org.rul.nutrimake.model.projection.NombreClientes


/**
 * Created by Rul on 11/10/2017.
 */
@Dao
interface ClienteDao {
    @get:Query("SELECT * FROM cliente")
    val all: List<Cliente>

    @Query("SELECT * FROM cliente WHERE uid IN (:clienteIds)")
    fun loadAllByIds(clienteIds: IntArray): List<Cliente>

    @Query("SELECT * FROM cliente WHERE nombre LIKE :pNombre AND " + " apellidos LIKE :pApellidos LIMIT 1")
    fun findByName(pNombre: String, pApellidos: String?): Cliente

    //LiveData ejemplo
    @Query("SELECT nombre, apellidos FROM cliente")
    fun loadFullName(): LiveData<List<NombreClientes>>

    //RxJava2 ejemplo
    @Query("SELECT * from cliente where id = :id LIMIT 1")
    fun loadClienteById(id: Int): Flowable<Cliente>

    @Query("SELECT * FROM cliente " +
            " INNER JOIN analitica ON analitica.cliente_id = cliente.id LIMIT 5")
    fun loadRawClienteConAnalitica(minAge: Int): Cursor

    @Insert
    fun insertAll(vararg clientes: Cliente)

    @Delete
    fun delete(cliente: Cliente)
}
